src/engine/Join.h

```diff
@@ -94,6 +94,18 @@ class Join : public Operation {
   void join(const IdTable& a, ColumnIndex jc1, const IdTable& b,
             ColumnIndex jc2, IdTable* result) const;
 
+ private:
+  // Part of the implementation of `createResult`. This function is called when
+  // the result should be yielded lazily.
+  // The semantics of action can be seen as
+  // runJoinAndReturnFinalResult(callbackForIntermediateResults).
```

Comment:
more precisely it is the callback for the `CombinedRowAdder`. 
And that last line can be rewritten in full sentences.
And the comment should state that this function turns the callback-based approach of lazily yielding intermediate results via the `AddCombinedRowAdder` into a `generator` using an explicit thread.

With these informations, one can understand the code.